### PR TITLE
Fix/cobra0.6

### PR DESCRIPTION
**Pull request by hredestig**
Remove all cobrapy subclasses and make cameo cobrapy06 compatible. Huge pull-request but most changes are straight-forward and rather trivial. When reviewing, suggest to take one commit at at time as the message indicates what was done in isolation. All changes up for discussion of course! Note that this is PR to devel06 so we keep devel 'steady as she goes' for now.

### CODE REVIEW

File: cameo/strain_design/heuristic/evolutionary_based.py [hunk 1]
```diff
@@ -478,17 +478,17 @@ def _process_solutions(self):
             self._processed_solutions = processed_solutions
 
     def display_on_map(self, index=0, map_name=None, palette="YlGnBu"):
-        with TimeMachine() as tm:
+        with self._model:
             for ko in self.data_frame.loc[index, "reactions"]:
-                self._model.reactions.get_by_id(ko).swap_cofactors(self._swap_pairs, tm)
+                self._model.reactions.get_by_id(ko).swap_cofactors(self._swap_pairs)
```
Comment (joaocardoso) [inline cameo/strain_design/heuristic/evolutionary_based.py]: Reaction.swap_cofactors do not exist anymore. You can use `cameo.core.manipulation.swap_cofactors`.

File: cameo/strain_design/heuristic/evolutionary/processing.py [hunk 2]
```diff
@@ -95,17 +95,16 @@ def process_gene_knockout_solution(model, solution, simulation_method, simulatio
         A list with: reactions, genes, size, fva_min, fva_max, target flux, biomass flux, yield, fitness
     """
 
-    with TimeMachine() as tm:
+    with model:
```
Comment (joaocardoso) [inline cameo/strain_design/heuristic/evolutionary/processing.py]: The other method bellow (process_reaction_swap_solution) needs to be replaced too.

File: cameo/strain_design/deterministic/linear_programming.py [hunk 3]
```diff
@@ -269,9 +269,7 @@ def run(self, max_knockouts=5, biomass=None, target=None, max_results=1, *args,
 
                 if len(knockouts) < max_knockouts:
                     self._number_of_knockouts_constraint.lb = self._number_of_knockouts_constraint.ub - len(knockouts)
-
-                tm(do=partial(self._model.solver.add, integer_cut),
-                   undo=partial(self._model.solver.remove, integer_cut))
+                self._model.add_cons_vars(integer_cut)
```
Comment (joaocardoso) [inline cameo/strain_design/deterministic/linear_programming.py]: shouldn't we use `add` and `remove` only and let the solver handle the adding of constraints and variables.

File: cameo/flux_analysis/simulation.py [hunk 4]
```diff
@@ -75,10 +73,9 @@ def fba(model, objective=None, reactions=None, *args, **kwargs):
...
 
     """
-    with TimeMachine() as tm:
+    with model:
         if objective is not None:
-            tm(do=partial(setattr, model, 'objective', objective),
-               undo=partial(setattr, model, 'objective', model.objective))
+            model.objective = objective
         model.solver.optimize()
         if model.solver.status != 'optimal':
```
Comment (joaocardoso) [inline cameo/flux_analysis/simulation.py]: I think it should raise the correct status: Infeasible, Unbounded, etc. We have some dictionaries inplace for this.
The type of exception is informative and necessary to decide how to handle the resolution.

File: cameo/strain_design/heuristic/evolutionary_based.py [hunk 5]
```diff
@@ -479,17 +478,17 @@ def _process_solutions(self):
...
             fluxes = self._simulation_method(self._model, **self._simulation_kwargs)
             fluxes.display_on_map(map_name=map_name, palette=palette)
 
     def plot(self, index=0, grid=None, width=None, height=None, title=None, palette=None, **kwargs):
         wt_production = phenotypic_phase_plane(self._model, objective=self._target, variables=[self._biomass])
-        with TimeMachine() as tm:
+        with self._model:
             for ko in self.data_frame.loc[index, "reactions"]:
-                self._model.reactions.get_by_id(ko).swap_cofactors(self._swap_pairs, tm)
+                self._model.reactions.get_by_id(ko).swap_cofactors(self._swap_pairs)
```
Comment (joaocardoso) [inline cameo/strain_design/heuristic/evolutionary_based.py]: same as before

File: cameo/flux_analysis/simulation.py [hunk 6]
```diff
@@ -226,7 +228,7 @@ def create_objective(model, variables):
 
         cache.add_objective(create_objective, None, cache.variables.values())
 
-        solution = model.solve()
+        solution = model.optimize()
```
Comment (joaocardoso) [inline cameo/flux_analysis/simulation.py]: Shouldn't it catch non-optimal solutions here too?

File: cameo/flux_analysis/simulation.py [hunk 7]
```diff
@@ -418,7 +420,7 @@ def update_lower_constraint(model, constraint, reaction, variable, flux_value, e
         model.objective = model.solver.interface.Objective(add([mul([One, var]) for var in cache.variables.values()]),
                                                            direction='min')
         try:
-            solution = model.solve()
+            solution = model.optimize()
```
Comment (joaocardoso) [inline cameo/flux_analysis/simulation.py]: and here too.

File: cameo/flux_analysis/simulation.py [hunk 8]
```diff
@@ -74,59 +73,61 @@ def fba(model, objective=None, reactions=None, *args, **kwargs):
...
     """
-    with TimeMachine() as tm:
+    with model:
         if objective is not None:
-            tm(do=partial(setattr, model, 'objective', objective),
-               undo=partial(setattr, model, 'objective', model.objective))
-        solution = model.solve()
+            model.objective = objective
+        model.solver.optimize()
+        if model.solver.status != 'optimal':
```
Comment (joaocardoso) [inline cameo/flux_analysis/simulation.py]: I think it should raise an exception according to the status: infeasible, unbounded, etc.

File: cameo/strain_design/heuristic/evolutionary_based.py (hunk 1 above)
Comment (hredestig) [inline cameo/strain_design/heuristic/evolutionary_based.py]: Fixed

File: cameo/strain_design/heuristic/evolutionary/processing.py (hunk 2 above)
Comment (hredestig) [inline cameo/strain_design/heuristic/evolutionary/processing.py]: Fixed

File: cameo/strain_design/deterministic/linear_programming.py (hunk 3 above)
Comment (hredestig) [inline cameo/strain_design/deterministic/linear_programming.py]: `model.add_cons_vars` is context-aware so handles reverting as well.

File: cameo/flux_analysis/simulation.py (hunk 4 above)
Comment (hredestig) [inline cameo/flux_analysis/simulation.py]: Agree, this was a bit messy still and now fixed in cobrapy and here as well

File: cameo/flux_analysis/simulation.py (hunk 8 above)
Comment (hredestig) [inline cameo/flux_analysis/simulation.py]: dito

File: cameo/flux_analysis/simulation.py (hunk 6 above)
Comment (hredestig) [inline cameo/flux_analysis/simulation.py]: Maybe, this is just a replacement for the now removed `model.solve`. Error handling there is scheduled to change so will be updated here when that happens as well.

File: cameo/flux_analysis/simulation.py (hunk 7 above)
Comment (hredestig) [inline cameo/flux_analysis/simulation.py]: same as above

File: cameo/strain_design/deterministic/linear_programming.py (hunk 3 above)
Comment (joaocardoso) [inline cameo/strain_design/deterministic/linear_programming.py]: Alright. BTW, what does `add_cons_vars` stands for? Add constraints and vars? It is not that obvious I think.

**Comment by joaocardoso** [comment]
Looks great! I have general question: what happened to the tests? I cannot see travis build results here.

**Comment by hredestig** [comment]
A good point, forgot to turn that on, done!

File: cameo/strain_design/deterministic/linear_programming.py (hunk 3 above)
Comment (hredestig) [inline cameo/strain_design/deterministic/linear_programming.py]: I know.. :/ but `add_constraints_or_variables` is too long..

Comment (phantomas1234) [inline cameo/strain_design/deterministic/linear_programming.py]: I would argue that `add_constraints_or_variables` is not too long if you're using a proper editor

Comment (hredestig) [inline cameo/strain_design/deterministic/linear_programming.py]: you're about to trigger a line length discussion, sure you want to go there..? :)

Comment (phantomas1234) [inline cameo/strain_design/deterministic/linear_programming.py]: line length is configured to 120 characters for cameo https://github.com/biosustain/cameo/blob/devel/setup.cfg#L15 so no issue there :-)

File: tests/data/update-pickles.py [hunk 9]
```diff
@@ -1,13 +1,15 @@
...
 import cobra.test
+import optlang
 
 from cameo import load_model
 
-ijo = load_model('iJO1366.xml')
+ijo = load_model('iJO1366.xml', solver_interface=optlang.glpk_interface)
 with open('iJO1366.pickle', 'wb') as out:
-    pickle.dump(ijo, out)
+    pickle.dump(ijo, out, protocol=2)
```
Comment (Midnighter) [inline tests/data/update-pickles.py]: I prefer using `pickle.HIGHEST_PROTOCOL` although that has equalled two for the past decade or so.

**Pull request merged**